src/Functions/FunctionStringCompare.cpp

```diff
@@ -0,0 +1,446 @@
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/IDataType.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/FunctionsComparison.h>
+#include <Functions/IFunction.h>
+#include <Interpreters/Context.h>
+#include <Common/memcmpSmall.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_COLUMN;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+// usage:
+// stringCompare(s, t)
+//  Compare s with t.
+// stringCompare(s, t, offset_s, offset_t, n)
+//  Compare s from offset_s to offset_s + n with t from offset_t to offset_t + n.
+class FunctionStringCompare : public IFunction
+{
+public:
+    static constexpr auto name = "stringCompare";
+    static FunctionPtr create(const ContextPtr /*context*/) { return std::make_shared<FunctionStringCompare>(); }
+
+    String getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 0; }
+
+    bool isVariadic() const override { return true; }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        if (arguments.size() != 2 && arguments.size() != 5)
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function {} requires 2 or 5 arguments", getName());
+
+        FunctionArgumentDescriptors mandatory_args{
+            {"str1", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isStringOrFixedString), nullptr, "String"},
+            {"str2", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isStringOrFixedString), nullptr, "String"},
+        };
+        FunctionArgumentDescriptors optional_args{
+            {"str1_offset",
+             static_cast<FunctionArgumentDescriptor::TypeValidator>(&isUInt),
+             static_cast<FunctionArgumentDescriptor::ColumnValidator>(&FunctionStringCompare::isConstColumn),
+             "UInt"},
+            {"str2_offset",
+             static_cast<FunctionArgumentDescriptor::TypeValidator>(&isUInt),
+             static_cast<FunctionArgumentDescriptor::ColumnValidator>(&FunctionStringCompare::isConstColumn),
+             "UInt"},
+            {"n",
+             static_cast<FunctionArgumentDescriptor::TypeValidator>(&isUInt),
+             static_cast<FunctionArgumentDescriptor::ColumnValidator>(&FunctionStringCompare::isConstColumn),
+             "UInt"},
+        };
+        validateFunctionArguments(*this, arguments, mandatory_args, optional_args);
+
+        return std::make_shared<DataTypeInt8>();
+    }
+
+    DataTypePtr getReturnTypeForDefaultImplementationForDynamic() const override { return std::make_shared<DataTypeInt8>(); }
+
+    ColumnPtr
+    executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & /*result_type*/, size_t /*input_rows_count*/) const override
```

Comment:
The code should assume that `input_rows_count` is the number of rows in the input chunk (instead of infering that from the chunk itself).